crates/runtime/src/memory.rs

```diff
@@ -389,26 +407,34 @@ impl Memory {
     pub unsafe fn grow(
         &mut self,
         delta_pages: u64,
-        limiter: Option<&mut dyn ResourceLimiter>,
+        mut limiter: Option<&mut dyn ResourceLimiter>,
     ) -> Option<usize> {
         let old_byte_size = self.byte_size();
+        // Wasm spec: when growing by 0 pages, always return the current size.
         if delta_pages == 0 {
             return Some(old_byte_size);
         }
 
+        // If any of the conversions from delta_pages to new_byte_size overflows,
+        // new_byte_size will be usize::MAX, an amount of memory which is never
+        // possible to allocate, but is still
         let new_byte_size = usize::try_from(delta_pages)
-            .ok()?
-            .checked_mul(WASM_PAGE_SIZE)?
-            .checked_add(old_byte_size)?;
+            .unwrap_or(usize::MAX)
+            .checked_mul(WASM_PAGE_SIZE)
+            .unwrap_or(usize::MAX)
+            .checked_add(old_byte_size)
+            .unwrap_or(usize::MAX);
```

Comment:
Along the lines of my comment up above, I think actually these `usize::MAX` values might be good to replace with `absolute_max` as calculated in the `new` function. That way `usize::MAX` doesn't have to get specifically tested for.